activerecord/lib/arel/visitors/to_sql.rb

```diff
@@ -578,9 +578,41 @@ def visit_Arel_Nodes_And(o, collector)
         end
 
         def visit_Arel_Nodes_Or(o, collector)
-          collector = visit o.left, collector
-          collector << " OR "
-          visit o.right, collector
+          tree = {}.compare_by_identity
+          done = nil
+
+          while o
+            if done&.equal?(o.right)
+              done = o
+              o = tree[o]
+              next
+            end
+
+            unless done&.equal?(o.left)
+              if o.left.is_a?(Arel::Nodes::Or)
+                tree[o.left] = o
+                o = o.left
+                next
+              end
+
+              visit o.left, collector
+            end
+
+            collector << " OR "
+
+            if o.right.is_a?(Arel::Nodes::Or)
+              tree[o.right] = o
+              o = o.right
+              next
+            end
+
+            visit o.right, collector
+
+            done = o
+            o = tree[o]
+          end
```

Comment:
Would this work as well?

```suggestion
          stack = [o]

          while (o = stack.pop)
            if o.is_a?(Arel::Nodes::Or)
              stack << o.right
              stack << o.left
            else
              visit o, collector
              collector << " OR " unless stack.empty?
            end
          end
```

Reply:
👍 I've got the same feedback to use stack from a friend.